modules/python/python-types.c

```diff
@@ -0,0 +1,446 @@
+is_py_obj_bytes_or_string_type(PyObject *obj)
+{
+  return PyBytes_CheckExact(obj) || PyUnicode_CheckExact(obj);
+}
+
+/* NOTE: This function returns a managed memory area pointing to an utf8
+ *       representation of the string, with the following constraints:
+ *
+ *   1) We basically assume that non-unicode strings are in utf8 or at least
+ *      utf8 compatible (e.g. ascii).  It doesn't really make sense otherwise.
+ *      If we don't the resulting string is not going to be utf8, rather it
+ *      would be the system codepage.
+ *
+ *   2) We are using the utf8 cache in the unicode instance.
+ */
+gboolean
+py_bytes_or_string_to_string(PyObject *obj, const gchar **string)
+{
+  if (!is_py_obj_bytes_or_string_type(obj))
+    {
+      _log_py_object_conversion_error_with_invalid_type(obj, "str/bytes");
+      return FALSE;
+    }
+
+  const gchar *string_local;
+
+  if (PyBytes_Check(obj))
+    {
+      string_local = PyBytes_AsString(obj);
+    }
+  else if (PyUnicode_Check(obj))
+    {
+      string_local = PyUnicode_AsUTF8(obj);
+    }
+  else
+    {
+      msg_error("Unexpected python string value");
+      return FALSE;
+    }
+
+  if (!string_local)
+    {
+      _log_py_object_conversion_error_with_exception();
+      return FALSE;
+    }
+
+  *string = string_local;
+  return TRUE;
+}
+
+gboolean
+py_long_to_long(PyObject *obj, gint64 *l)
+{
+  if (!PyLong_Check(obj))
+    {
+      _log_py_object_conversion_error_with_invalid_type(obj, "int");
+      return FALSE;
+    }
+
+  gint64 l_local = PyLong_AsLong(obj);
+  if (PyErr_Occurred())
+    {
+      _log_py_object_conversion_error_with_exception();
+      return FALSE;
+    }
+
+  *l = l_local;
+  return TRUE;
+}
+
+gboolean
+py_double_to_double(PyObject *obj, gdouble *d)
+{
+  if (!PyFloat_Check(obj))
+    {
+      _log_py_object_conversion_error_with_invalid_type(obj, "float");
+      return FALSE;
+    }
+
+  gdouble d_local = PyFloat_AsDouble(obj);
+  if (PyErr_Occurred())
+    {
+      _log_py_object_conversion_error_with_exception();
+      return FALSE;
+    }
+
+  *d = d_local;
+  return TRUE;
+}
+
+gboolean
+py_boolean_to_boolean(PyObject *obj, gboolean *b)
+{
+  if (!PyBool_Check(obj))
+    {
+      _log_py_object_conversion_error_with_invalid_type(obj, "bool");
+      return FALSE;
+    }
+
+  if (obj == Py_True)
+    {
+      *b = TRUE;
+      return TRUE;
+    }
+
+  if (obj == Py_False)
+    {
+      *b = FALSE;
+      return TRUE;
+    }
+
+  return FALSE;
+}
+
+gboolean
+py_list_to_list(PyObject *obj, gchar **list)
+{
+  if (!PyList_Check(obj))
+    {
+      _log_py_object_conversion_error_with_invalid_type(obj, "list");
+      return FALSE;
+    }
+
+  ScratchBuffersMarker marker;
+  GString *buffer = scratch_buffers_alloc_and_mark(&marker);
+
+  for (Py_ssize_t i = 0; i < PyList_GET_SIZE(obj); i++)
+    {
+      PyObject *element = PyList_GET_ITEM(obj, i);
+
+      const gchar *element_as_str;
+      if (!py_bytes_or_string_to_string(element, &element_as_str))
```

Comment:
The py_bytes_or_string_to_string() function checks if the argument is a string and returns FALSE in case it isn't.

So non-strings in the list should be reported correctly.